linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/rm/external/yarn/YarnResourceRequester.java

```diff
@@ -0,0 +1,426 @@
+        if (yarnQueueName.equals(realQueueName)) {
+          return Optional.of(q);
+        } else if (realQueueName.startsWith(yarnQueueName + ".")) {
+          Optional childQueue = getQueue(getChildQueues(q), realQueueName);
+          if (childQueue.isPresent()) {
+            return childQueue;
+          }
+        }
+      }
+      return Optional.empty();
+    } else if (queues instanceof ObjectNode) {
+      ObjectNode queueObj = (ObjectNode) queues;
+      JsonNode queueName = queueObj.get("queueName");
+      if (queueName != null && queueName.asText().equals(realQueueName)) {
+        return Optional.of(queueObj);
+      } else {
+        JsonNode childQueues = queueObj.get("childQueues");
+        if (childQueues == null) {
+          return Optional.empty();
+        } else {
+          return getQueue(childQueues, realQueueName);
+        }
+      }
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  public static JsonNode getChildQueues(JsonNode resp) {
+    JsonNode queues = resp.get("childQueues").get("queue");
+    if (queues != null && !queues.isNull() && queues.size() > 0) {
+      logger.info("queues:" + queues);
+      return queues;
+    } else {
+      return resp.get("childQueues");
+    }
+  }
+
+  public Optional<JsonNode> getQueueOfCapacity(JsonNode queues, String realQueueName) {
+    if (queues.isArray()) {
+      for (JsonNode q : queues) {
+        String yarnQueueName = q.get("queueName").asText();
+        if (yarnQueueName.equals(realQueueName)) {
+          return Optional.of(q);
+        } else if (q.has("queues")) {
+          Optional<JsonNode> matchQueue = getQueueOfCapacity(q.get("queues"), realQueueName);
+          if (matchQueue.isPresent()) {
+            return matchQueue;
+          }
+        }
+      }
+      return Optional.empty();
+    } else if (queues.isObject()) {
+      if (queues.has("queueName") && queues.get("queueName").asText().equals(realQueueName)) {
+        return Optional.of(queues);
+      } else if (queues.has("queues")) {
+        Optional<JsonNode> matchQueue = getQueueOfCapacity(queues.get("queues"), realQueueName);
+        if (matchQueue.isPresent()) {
+          return matchQueue;
+        }
+      }
+      return Optional.empty();
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  static JsonNode getChildQueuesOfCapacity(JsonNode resp) {
+    // = resp \ "queues" \ "queue"
+    return resp.path("queues").path("queue");
+  }
+
+  public Pair<YarnResource, YarnResource> getResources(
+      String rmWebAddress, String realQueueName, String queueName) {
+    JsonNode resp = getResponseByUrl("scheduler", rmWebAddress);
+    String schedulerType = resp.path("scheduler").path("schedulerInfo").path("type").asText();
+    if ("capacityScheduler".equals(schedulerType)) {
+      realQueueName = queueName;
+      JsonNode childQueues = getChildQueuesOfCapacity(resp.path("scheduler").path("schedulerInfo"));
```

Comment:
can we parse `resp.path("scheduler").path("schedulerInfo")` once and used latter ?

Reply:
Yeah , thanks for your good advice. I have modified it .